website/app/routes/2024-10-05-hackathon-at-sentry.tsx

```diff
@@ -2,15 +2,17 @@ import { SentryLogoIcon } from '~/modules/components/ui/icons';
 import { EventDetailsPage, PhotosSection } from '~/modules/event-details/components';
 import { Section } from '~/modules/components/ui/section';
 import { meta } from '~/modules/event-details/meta';
-import { eventDetailsLoader } from '~/modules/event-details/loader.sever';
 import { useLoaderData } from '@remix-run/react';
+import { json, LoaderFunctionArgs } from '@remix-run/node';
 
 export { headers } from '~/modules/header.server';
 
 export { meta };
 
-export function loader() {
-  return eventDetailsLoader('2024-10-05-hackathon-at-sentry');
+export async function loader({ context }: LoaderFunctionArgs) {
```

Comment:
Okay, this is obviously where all the legwork pays off. It's more lines but very nice to read and I can see how this would pay off in the `action` and `loader` functions when they get more complex. But I am also a bit more afraid of debugging this in case something goes wrong because it's not purely a plain function that is being called. It's a bunch of layers configured elsewhere, influenced by the global context.

Reply:
well that would/should exclusively be the `handler` which is a plain function.
yes you have the middleware too, but most likely the handler.

here one can argue that debugging is more complex cause instances of anything can come from anywhere via import.

Reply:
Exactly, and that's precisely the point where I am like, this is probably overkill...

Reply:
This & my concern that middlewares now can't really be handler specific, right?

From previous comment: "For instance, how would you solve: 'this handler data should never be cached' or 'this handler, I want to log more than just the default' use cases in the current implementation?"

If things like caching aren't middleware but plain function calls that return stuff, then you can provide the required arguments in each `loader` & `action` function to alter the caching time or logging logic. Each `loader` and `action` function contains all the logic, all the function calls, and can be managed independently.

Reply:
First here loader/action are framework-specific, so importing remix-specific stuff makes it "not decoupled" enough.
Loader/action are controllers, the glue between the framework and the domain/infra.

And I think it's a wrong approach to think about it this way.  Per definition, a middleware is not "handler specific" there is no point having that. if you try to do that, there is a flaw somewhere.

Also, you don't need a middleware all the time, you can still manage anything you want in the handler and have logic in the handlers. (like you do today at the end)

Also, you can disable the cache per `query/command` for instance (already possible in missive.js)

If you need different cache storage per `query/command` (assuming it would make sense...based on a different discussion we 've got) you could have 2 cacher middlewares 1 for redis and 1 for in memory, both using the built-in middleware but with 2 different adapters and they would have each "per intent" configuration.

also I could see a world where you wrap the dispatch in cachified for instance, if you prefer. A bit weird but not invalid. 

Missive.js is not a dropped-in replacement for everything it's a lib that helps implement the paradigm with a bus CQRS. if you have 246 queries/ commands, I am pretty confident it would bring more value that it would remove.

In short the goal, on any projects, the way I envisioned them, is:
- to have contracts/interfaces you share across the app disconnected from any implementations
- to have domain(s) folders/files that rely ONLY on the contracts/interfaces
- to have infrastructure(s) folders/files that are the actual implementations (respecting the interfaces)

That is before any Bus, CQRS or, anything.
Then Missive.js brings CQRS, a paradigm on top that works well with the above approach.

The `modules` approach works well, it's not bad, but it does not scale the same way it's actually bloated when you think about it, not easily testable, not easily interchangeable, not yelling what it does but more what tool it is, etc.

It's just another approach, I do think hexagonal architecture is better suited for big professional projects, and yes overkill for small projects like this one.


my 2 cents ;) I have implemented hundreds of projects some with CQRS some without it. Like any tech, we need to use it when it makes sense ;-)

=> here, this PR is on a small project and was to show/share another way.